File: disciple/win32/win32_scrollbar.ml
(* OCaml-Win32
 * win32_scrollbar.ml
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *)

open Win32_core

(********************************************************************)
(* User Interface Services: Controls: Scroll Bars *)

external enable_scroll_bar : wnd:hwnd -> typ:scroll_bar_type -> option:enable_scroll_bar_option -> bool = "enable_scroll_bar"

external get_scroll_info : wnd:hwnd -> typ:scroll_bar_type -> get_scroll_info = "get_scroll_info"

external get_scroll_pos : wnd:hwnd -> typ:scroll_bar_type -> int = "get_scroll_pos"

external get_scroll_range : wnd:hwnd -> typ:scroll_bar_type -> (int * int) = "get_scroll_range"

external scroll_dc : dc:hdc -> amount:point -> scroll:rect -> clip:rect -> update:hrgn -> rect = "scroll_dc"

external scroll_window :
    wnd:hwnd ->
    amount:point ->
    scroll:rect ->
    clip:rect ->
    update:hrgn ->
    options:scroll_window_option list ->
    rect = "scroll_window_bytecode" "scroll_window_native"

external set_scroll_info : wnd:hwnd -> typ:scroll_bar_type -> info:set_scroll_info -> redraw:bool -> int = "set_scroll_info"

external set_scroll_pos : wnd:hwnd -> typ:scroll_bar_type -> pos:int -> redraw:bool -> int = "set_scroll_pos"

external set_scroll_range : wnd:hwnd -> typ:scroll_bar_type -> (int * int) -> redraw:bool -> unit = "set_scroll_range"

external show_scroll_bar : wnd:hwnd -> typ:scroll_bar_type -> show:bool -> unit = "show_scroll_bar"

(* Styles *)

type scrollbar_style_option =
    SBS_BOTTOMALIGN
  | SBS_HORZ
  | SBS_LEFTALIGN
  | SBS_RIGHTALIGN
  | SBS_SIZEBOX
  | SBS_SIZEBOXBOTTOMRIGHTALIGN
  | SBS_SIZEBOXTOPLEFTALIGN
  | SBS_SIZEGRIP
  | SBS_TOPALIGN
  | SBS_VERT

external scrollbar_style_of_option : scrollbar_style_option -> style = "scrollbar_style_of_option"

let scrollbar_style_of_options = special_style_of_options ~special_func:scrollbar_style_of_option

(* Messages *)

let sbm_enable_arrows = 0x00E4
external pack_sbm_enable_arrows : option:enable_scroll_bar_option -> message_contents = "pack_sbm_enable_arrows"
external unpack_sbm_enable_arrows : message_result -> unit = "unpack_sbm_enable_arrows"
let issue_sbm_enable_arrows ~sender ~option = unpack_sbm_enable_arrows (sender ~msg:(pack_sbm_enable_arrows ~option))

let sbm_getpos = 0x00E1
let pack_sbm_getpos = pack_simple_message ~id:sbm_getpos
external unpack_sbm_getpos : message_result -> int = "unpack_sbm_getpos"
let issue_sbm_getpos ~sender = unpack_sbm_getpos (sender ~msg:pack_sbm_getpos)

let sbm_getrange = 0x00E3
external issue_sbm_getrange : sender:message_sender -> (int * int) = "issue_sbm_getrange"

let sbm_getscrollinfo = 0x00EA
external issue_sbm_getscrollinfo : sender:message_sender -> get_scroll_info = "issue_sbm_getscrollinfo"

let sbm_setpos = 0x00E0
external pack_sbm_setpos : pos:int -> redraw:bool -> message_contents = "pack_sbm_setpos"
external unpack_sbm_setpos : message_result -> int = "unpack_sbm_setpos"
let issue_sbm_setpos ~sender ~pos ~redraw = unpack_sbm_setpos (sender ~msg:(pack_sbm_setpos ~pos ~redraw))

let sbm_setrange = 0x00E2
external pack_sbm_setrange : range:(int * int) -> message_contents = "pack_sbm_setrange"
external unpack_sbm_setrange : message_result -> int = "unpack_sbm_setrange"
let issue_sbm_setrange ~sender ~range = unpack_sbm_setrange (sender ~msg:(pack_sbm_setrange ~range))

let sbm_setrangeredraw = 0x00E6
external pack_sbm_setrangeredraw : range:(int * int) -> message_contents = "pack_sbm_setrangeredraw"
external unpack_sbm_setrangeredraw : message_result -> int = "unpack_sbm_setrangeredraw"
let issue_sbm_setrangeredraw ~sender ~range = unpack_sbm_setrangeredraw (sender ~msg:(pack_sbm_setrangeredraw ~range))

let sbm_setscrollinfo = 0x00E9
external issue_sbm_setscrollinfo : sender:message_sender -> info:set_scroll_info -> redraw:bool -> int = "issue_sbm_setscrollinfo"

let wm_ctlcolorscrollbar = 0x0137
let on_wm_ctlcolorscrollbar = on_ctlcolor_message ~id:wm_ctlcolorscrollbar
let _ = add_explicit_dialog_return wm_ctlcolorscrollbar

type wm_scroll_handler =
    wnd:hwnd ->
    msg:message_contents ->
    action:scroll_action ->
    bar:hwnd ->
    unit message_handler_result
external deliver_wm_scroll :
    handler:wm_scroll_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    unit message_handler_result = "deliver_wm_scroll"

let wm_hscroll = 0x0114
let on_wm_hscroll = on_message ~id:wm_hscroll ~deliver:deliver_wm_scroll ~return:return_simple_message

let wm_vscroll = 0x0115
let on_wm_vscroll = on_message ~id:wm_vscroll ~deliver:deliver_wm_scroll ~return:return_simple_message

File: disciple/win32/win32_scrollbar.c
/* OCaml-Win32
 * win32_scrollbar.c
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "win32.h"

/********************************************************************/
/* User Interface Services: Controls: Scroll Bars */

CAMLprim value enable_scroll_bar(value wnd, value type, value option)
{
    CAMLparam3(wnd, type, option);
    HWND hwnd;
    BOOL result;

    hwnd = Handle_val(wnd);
    enter_blocking_section();
    result = EnableScrollBar(
        hwnd,
        Constant_val(type, scroll_bar_types),
        Constant_val(option, enable_scroll_bar_options));
    leave_blocking_section();
    CAMLreturn(Val_bool(result));
}

CAMLprim value get_scroll_info(value wnd, value type)
{
    SCROLLINFO si;

    si.cbSize = sizeof(SCROLLINFO);
    si.fMask = SIF_ALL;
    if (!GetScrollInfo(Handle_val(wnd), Constant_val(type, scroll_bar_types), &si))
        raise_last_error();
    return alloc_getscrollinfo(&si);
}

CAMLprim value get_scroll_pos(value wnd, value type)
{
    return Val_int(GetScrollPos(Handle_val(wnd), Constant_val(type, scroll_bar_types)));
}

CAMLprim value get_scroll_range(value wnd, value type)
{
    CAMLparam2(wnd, type);
    int min, max;
    CAMLlocal1(val);

    if (!GetScrollRange(Handle_val(wnd), Constant_val(type, scroll_bar_types), &min, &max))
        raise_last_error();
    val = alloc_tuple(2);
    Store_field(val, 0, Val_int(min));
    Store_field(val, 1, Val_int(max));
    CAMLreturn(val);
}

CAMLprim value scroll_dc(value hdc, value amount, value scroll, value clip, value update)
{
    RECT scroll_rect, clip_rect, update_rect;
    BOOL result;

    rect_val(scroll, &scroll_rect);
    rect_val(clip, &clip_rect);
    result = ScrollDC(
        Handle_val(hdc),
        X(amount), Y(amount),
        &scroll_rect,
        &clip_rect,
        Handle_val(update),
        &update_rect);
    if (!result)
        raise_last_error();
    return alloc_rect(&update_rect);
}

CAMLprim value scroll_window_native(value wnd, value amount, value scroll, value clip, value update, value options)
{
    CAMLparam5(wnd, amount, scroll, clip, update);
    CAMLxparam1(options);
    HWND hwnd;
    POINT amt;
    RECT scroll_rect, clip_rect;
    HRGN upd;
    RECT update_rect;
    UINT opt;
    int result;
    CAMLlocal1(val);

    hwnd = Handle_val(wnd);
    amt = point_val(amount);
    rect_val(scroll, &scroll_rect);
    rect_val(clip, &clip_rect);
    upd = Handle_val(update);
    opt = flags_val(options, scroll_window_options);
    enter_blocking_section();
    result = ScrollWindowEx(
        hwnd,
        amt.x, amt.y,
        &scroll_rect,
        &clip_rect,
        upd,
        &update_rect,
        opt);
    leave_blocking_section();
    if (result == ERROR)
        raise_last_error();
    val = alloc_rect(&update_rect);
    CAMLreturn(val);
}

CAMLprim value scroll_window_bytecode(value* argv, int argn)
{
    return scroll_window_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

CAMLprim value set_scroll_info(value wnd, value type, value info, value redraw)
{
    CAMLparam4(wnd, type, info, redraw);
    HWND hwnd;
    SCROLLINFO si;
    int result;

    hwnd = Handle_val(wnd);
    setscrollinfo_val(info, &si);
    enter_blocking_section();
    result = SetScrollInfo(
        hwnd,
        Constant_val(type, scroll_bar_types),
        &si,
        Bool_val(redraw));
    leave_blocking_section();
    CAMLreturn(Val_int(result));
}

CAMLprim value set_scroll_pos(value wnd, value type, value pos, value redraw)
{
    CAMLparam4(wnd, type, pos, redraw);
    HWND hwnd;
    int result;

    hwnd = Handle_val(wnd);
    enter_blocking_section();
    result = SetScrollPos(
        hwnd,
        Constant_val(type, scroll_bar_types),
        Int_val(pos),
        Bool_val(redraw));
    leave_blocking_section();
    CAMLreturn(Val_int(result));
}

CAMLprim value set_scroll_range(value wnd, value type, value range, value redraw)
{
    CAMLparam4(wnd, type, range, redraw);
    HWND hwnd;
    BOOL result;

    hwnd = Handle_val(wnd);
    enter_blocking_section();
    result = SetScrollRange(
        hwnd,
        Constant_val(type, scroll_bar_types),
        Int_val(Field(range, 0)),
        Int_val(Field(range, 1)),
        Bool_val(redraw));
    leave_blocking_section();
    if (!result)
        raise_last_error();
    CAMLreturn(Val_unit);
}

CAMLprim value show_scroll_bar(value wnd, value type, value show)
{
    CAMLparam3(wnd, type, show);
    HWND hwnd;
    BOOL result;

    hwnd = Handle_val(wnd);
    enter_blocking_section();
    result = ShowScrollBar(hwnd, Constant_val(type, scroll_bar_types), Bool_val(show));
    leave_blocking_section();
    if (!result)
        raise_last_error();
    CAMLreturn(Val_unit);
}

/* Styles */

static DWORD scrollbar_style_options[] = {
    SBS_BOTTOMALIGN,
    SBS_HORZ,
    SBS_LEFTALIGN,
    SBS_RIGHTALIGN,
    SBS_SIZEBOX,
    SBS_SIZEBOXBOTTOMRIGHTALIGN,
    SBS_SIZEBOXTOPLEFTALIGN,
    SBS_SIZEGRIP,
    SBS_TOPALIGN,
    SBS_VERT
};
Define_static_constants(scrollbar_style_options);

CAMLprim value scrollbar_style_of_option(value option)
{
    return copy_int64((int64)Constant_val(option, scrollbar_style_options));
}

/* Messages */

CAMLprim value pack_sbm_enable_arrows(value option)
{
    return pack_message(SBM_ENABLE_ARROWS, Constant_val(option, enable_scroll_bar_options), 0);
}

CAMLprim value unpack_sbm_enable_arrows(value result)
{
    if (Int32_val(result) != TRUE)
        raise_last_error();
    return Val_unit;
}

CAMLprim value unpack_sbm_getpos(value result)
{
    return Val_int(Int32_val(result));
}

CAMLprim value issue_sbm_getrange(value sender)
{
    CAMLparam1(sender);
    int min, max;
    CAMLlocal2(contents, val);

    contents = pack_message(SBM_GETRANGE, (WPARAM)&min, (LPARAM)&max);
    callback(sender, contents);
    val = alloc_tuple(2);
    Store_field(val, 0, Val_int(min));
    Store_field(val, 1, Val_int(max));
    CAMLreturn(val);
}

CAMLprim value issue_sbm_getscrollinfo(value sender)
{
    CAMLparam1(sender);
    SCROLLINFO si;
    CAMLlocal3(contents, result, val);

    si.cbSize = sizeof(SCROLLINFO);
    si.fMask = SIF_ALL;
    contents = pack_message(SBM_GETSCROLLINFO, 0, (LPARAM)&si);
    result = callback(sender, contents);
    if (Int32_val(result) != TRUE)
        raise_last_error();
    val = alloc_getscrollinfo(&si);
    CAMLreturn(val);
}

CAMLprim value pack_sbm_setpos(value pos, value redraw)
{
    return pack_message(SBM_SETPOS, Int_val(pos), Bool_val(redraw));
}

CAMLprim value unpack_sbm_setpos(value result)
{
    return Val_int(Int32_val(result));
}

CAMLprim value pack_sbm_setrange(value range)
{
    return pack_message(SBM_SETRANGE, Int_val(Field(range, 0)), Int_val(Field(range, 1)));
}

CAMLprim value unpack_sbm_setrange(value result)
{
    return Val_int(Int32_val(result));
}

CAMLprim value pack_sbm_setrangeredraw(value range)
{
    return pack_message(SBM_SETRANGEREDRAW, Int_val(Field(range, 0)), Int_val(Field(range, 1)));
}

CAMLprim value unpack_sbm_setrangeredraw(value result)
{
    return Val_int(Int32_val(result));
}

CAMLprim value issue_sbm_setscrollinfo(value sender, value info, value redraw)
{
    CAMLparam3(sender, info, redraw);
    SCROLLINFO si;
    CAMLlocal2(contents, result);

    setscrollinfo_val(info, &si);
    contents = pack_message(SBM_SETSCROLLINFO, Bool_val(redraw), (LPARAM)&si);
    result = callback(sender, contents);
    CAMLreturn(Val_int(Int32_val(result)));
}

CAMLprim value deliver_wm_scroll(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    CAMLlocalN(args, 4);
    CAMLlocal1(result);

    args[0] = wnd;
    args[1] = msg;
    args[2] = alloc_scrollaction(LOWORD(Lparam(msg)), HIWORD(Lparam(msg)));
    args[3] = alloc_handle((HWND)Lparam(msg));
    result = callbackN(handler, 4, args);
    CAMLreturn(result);
}
